File: flutter-application/lib/models/recommendation_model.dart
class RecommendationModel {
  final String id;
  final String type;
  final String title;
  final String description;
  final String reason;
  final String action;
  final String priority;
  final String status;
  final bool isActive;
  final String? validUntil;
  final String? createdAt;

  final RecommendationMeta meta;
  final RecommendationDetails details;

  RecommendationModel({
    required this.id,
    required this.type,
    required this.title,
    required this.description,
    required this.reason,
    required this.action,
    required this.priority,
    required this.status,
    required this.isActive,
    required this.meta,
    required this.details,
    this.validUntil,
    this.createdAt,
  });

  factory RecommendationModel.fromJson(Map<String, dynamic> json) {
    return RecommendationModel(
      id: json['_id']?.toString() ?? '',
      type: json['type']?.toString() ?? '',
      title: json['title']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      reason: json['reason']?.toString() ?? '',
      action: json['action']?.toString() ?? '',
      priority: json['priority']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      isActive: json['isActive'] ?? false,
      validUntil: json['validUntil']?.toString(),
      createdAt: json['createdAt']?.toString(),
      meta: RecommendationMeta.fromJson(
        (json['meta'] as Map<String, dynamic>?) ?? {},
      ),
      details: RecommendationDetails.fromJson(
        (json['details'] as Map<String, dynamic>?) ?? {},
      ),
    );
  }
}

class RecommendationMeta {
  final int? soilMoisture;
  final String? soilType;
  final String? irrigationType;
  final String? growthStage;
  final String? weatherCondition;
  final num? temperature;
  final num? humidity;
  final String? scanId;
  final String? scanDisease;
  final num? scanConfidence;

  RecommendationMeta({
    this.soilMoisture,
    this.soilType,
    this.irrigationType,
    this.growthStage,
    this.weatherCondition,
    this.temperature,
    this.humidity,
    this.scanId,
    this.scanDisease,
    this.scanConfidence,
  });

  factory RecommendationMeta.fromJson(Map<String, dynamic> json) {
    return RecommendationMeta(
      soilMoisture: json['soilMoisture'] is num
          ? (json['soilMoisture'] as num).toInt()
          : null,
      soilType: json['soilType']?.toString(),
      irrigationType: json['irrigationType']?.toString(),
      growthStage: json['growthStage']?.toString(),
      weatherCondition: json['weatherCondition']?.toString(),
      temperature: json['temperature'] as num?,
      humidity: json['humidity'] as num?,
      scanId: json['scanId']?.toString(),
      scanDisease: json['scanDisease']?.toString(),
      scanConfidence: json['scanConfidence'] as num?,
    );
  }
}

class RecommendationDetails {
  final List<RecommendationScheduleItem> schedule;
  final List<String> notes;

  RecommendationDetails({
    required this.schedule,
    required this.notes,
  });

  factory RecommendationDetails.fromJson(Map<String, dynamic> json) {
    final rawSchedule = (json['schedule'] as List?) ?? [];
    final rawNotes = (json['notes'] as List?) ?? [];

    return RecommendationDetails(
      schedule: rawSchedule
          .map((e) => RecommendationScheduleItem.fromJson(
                (e as Map<String, dynamic>),
              ))
          .toList(),
      notes: rawNotes.map((e) => e.toString()).toList(),
    );
  }
}

class RecommendationScheduleItem {
  final String day;
  final String action;

  RecommendationScheduleItem({
    required this.day,
    required this.action,
  });

  factory RecommendationScheduleItem.fromJson(Map<String, dynamic> json) {
    return RecommendationScheduleItem(
      day: json['day']?.toString() ?? '',
      action: json['action']?.toString() ?? '',
    );
  }
}

File: backend/src/models/recommendation.model.js
import mongoose from "mongoose";

const recommendationScheduleSchema = new mongoose.Schema(
  {
    day: {
      type: String,
      required: true,
      trim: true,
    },
    action: {
      type: String,
      required: true,
      trim: true,
    },
  },
  { _id: false }
);

const recommendationDetailsSchema = new mongoose.Schema(
  {
    schedule: {
      type: [recommendationScheduleSchema],
      default: [],
    },
    notes: {
      type: [String],
      default: [],
    },
  },
  { _id: false }
);

const recommendationMetaSchema = new mongoose.Schema(
  {
    soilMoisture: { type: Number, default: null },
    soilType: { type: String, default: null },
    irrigationType: { type: String, default: null },
    growthStage: { type: String, default: null },
    weatherCondition: { type: String, default: null },
    temperature: { type: Number, default: null },
    humidity: { type: Number, default: null },
    scanId: { type: mongoose.Schema.Types.ObjectId, ref: "Scan", default: null },
    scanDisease: { type: String, default: null },
    scanConfidence: { type: Number, default: null },
  },
  { _id: false }
);

const recommendationSchema = new mongoose.Schema(
  {
    ownerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
      index: true,
    },

    farmId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Farm",
      required: true,
      index: true,
    },

    type: {
      type: String,
      enum: ["irrigation", "fertilization", "disease"],
      required: true,
      index: true,
    },

    title: {
      type: String,
      required: true,
      trim: true,
    },

    description: {
      type: String,
      required: true,
      trim: true,
    },

    reason: {
      type: String,
      required: true,
      trim: true,
    },

    action: {
      type: String,
      required: true,
      trim: true,
    },

    priority: {
      type: String,
      enum: ["high", "medium", "low"],
      default: "low",
      index: true,
    },

    status: {
      type: String,
      enum: ["pending", "done", "dismissed", "expired"],
      default: "pending",
      index: true,
    },

    isActive: {
      type: Boolean,
      default: true,
      index: true,
    },

    source: {
      type: [String],
      default: [],
    },

    meta: {
      type: recommendationMetaSchema,
      default: () => ({}),
    },

    details: {
      type: recommendationDetailsSchema,
      default: () => ({ schedule: [], notes: [] }),
    },

    validUntil: {
      type: Date,
      required: true,
      index: true,
    },

    lastTriggeredAt: {
      type: Date,
      default: Date.now,
    },

    resolvedAt: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

recommendationSchema.index({ ownerId: 1, status: 1, isActive: 1 });
recommendationSchema.index({ farmId: 1, type: 1, status: 1 });

const Recommendation = mongoose.model("Recommendation", recommendationSchema);

export default Recommendation;